clang/lib/Parse/ParseDeclCXX.cpp

```diff
@@ -4248,6 +4277,187 @@ ExceptionSpecificationType Parser::ParseDynamicExceptionSpecification(
+  } else {
+    if (!Tok.is(tok::r_paren))
+      SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
+    Cond = Actions.CreateRecoveryExpr(
+        Start, Start == Tok.getLocation() ? Start : PrevTokLocation, {},
+        Actions.getASTContext().BoolTy);
+  }
+
+  T.consumeClose();
+
+  if (CK == ContractKind::Pre) {
+    return Actions.ActOnPreContractAssert(KeywordLoc, Cond.get());
+  } else {
+    return Actions.ActOnPostContractAssert(KeywordLoc, Cond.get());
+  }
+}
+
+void Parser::ParsePostContract(Declarator &DeclaratorInfo) {
+  ConsumeToken();
+
+  ParseScope ParamScope(this, Scope::DeclScope | 
+                                  Scope::FunctionDeclarationScope |
+                                  Scope::FunctionPrototypeScope);
+
+  DeclaratorChunk::FunctionTypeInfo FTI = DeclaratorInfo.getFunctionTypeInfo();
+  for (unsigned i = 0; i != FTI.NumParams; ++i) {
+    ParmVarDecl *Param = cast<ParmVarDecl>(FTI.Params[i].Param);
+    Actions.ActOnReenterCXXMethodParameter(getCurScope(), Param);
+  }
+
+  if (Tok.isNot(tok::l_paren)) {
+    Diag(Tok.getLocation(), diag::err_expected) << tok::l_paren;
+    return;
+  }
+  ConsumeParen();
+
+  // Post contracts start with <identifier> colon <expression>
+  // As we have to support the "auto f() post (r : r > 42) {...}" case, we cannot parse here
+  // the return type is not guaranteed to be known until after the function body parses
+
+/*
+    if (Tok.isNot(tok::identifier)) {
+      Diag(Tok.getLocation(), diag::err_expected) << tok::identifier;
+      return;
+    }
+
+    ParsingDeclSpec DS(*this);
+
+    ParsedTemplateInfo TemplateInfo;
+    DeclSpecContext DSContext = getDeclSpecContextFromDeclaratorContext(DeclaratorContext::Block);
+    ParseDeclarationSpecifiers(DS, TemplateInfo, AS_none, DSContext);
+    
+    ParsedAttributes LocalAttrs(AttrFactory);
+    ParsingDeclarator D(*this, DS, LocalAttrs, DeclaratorContext::Block);
+
+    D.setObjectType(getAsFunction().getReturnType());
+    IdentifierInfo *Id = Tok.getIdentifierInfo();
+    SourceLocation IdLoc = ConsumeToken();
+    D.setIdentifier(Id, IdLoc);
+
+    Decl* ThisDecl = Actions.ActOnDeclarator(getCurScope(), D);
+    Actions.ActOnUninitializedDecl(ThisDecl);
+    Actions.FinalizeDeclaration(ThisDecl);
+    D.complete(ThisDecl);
+    if (Tok.isNot(tok::colon)) {
+      Diag(Tok.getLocation(), diag::err_expected) << tok::colon;
+      return;
+    }
+
+    ExprResult Expr = ParseExpression();
+    if (Expr.isInvalid()) {
+      Diag(Tok.getLocation(), diag::err_invalid_pcs);
+      return;
+    }
+    DeclaratorInfo.addPostContract(Expr.get());
+*/
+  ExprResult Expr = ParseExpression();
+  if (Expr.isInvalid()) {
+    Diag(Tok.getLocation(), diag::err_invalid_pcs);
+    return;
+  }
+  // DeclaratorInfo.addPostContract(Expr.get());
+
+  if (Tok.isNot(tok::r_paren)) {
+    Diag(Tok.getLocation(), diag::err_expected) << tok::r_paren;
+    return;
+  }
+  ConsumeParen();
+}
```

Comment:
Is that dead code?

Reply:
Yeah, but I still might want to reference it while building the result name introducer decl later.